### DISCUSSION TITLE
Block mode for @waline/vercel

### DISCUSSION THREAD

**Discussion by Mister-Hope** [walinejs/waline#792] [Ideas]
I am trying to find out a solution with #785. And  here is my possible solution:

1. Add a `TRACKIP` env variable and default to false.
1. add `blockedIPS` in options
1. Edit code to make the console prints all th ips in `X-Forwarded-For` header and the remote addr when `TRACKIP` has values, and also stores theme in a local log file. (With also the api path like `comment` `visitor`)
1. edit the admin panel to let admin users be able to view the log on line. This should probaly come along with the native 
1. Add a middleware beforeall services, if the request's `X-Forwarded-For` header or the send addr matches any of the `blockedIPS`, return 403 statuscode directly.

---

## Reason

 In most situations, the attackers should hold limited ips sending these request. (We are not facing large attacks like DDOS, if so , the user probably need a paid firewall).

The attacker's real ip should be one of the `X-Forwarded-For`  ips or the connect ip. And when under high frequent attacks, it's easy to find his real ip. Users can block his ip by adding the nearest unchanged ip on right  in  `X-Forwarded-For` header. That should block the attackers.

More aggresively, we can also have a `SECURE` env variable, which will record all the ip in  `X-Forwarded-For` header and block them with `IPQPS`, expect those in `permitedIPS`.

@lizheming  @ihackerx

**Comment by lizheming** [comment]
All request headers can be faked by user, including `X-Forwarded-For`. Recording is unmeaning if it can't be trusted.

If you watch the server code, you'll find server just get user's ip from `X-Forwarded-For` header. https://github.com/koajs/koa/blob/master/lib/request.js#L421-L459

**Reply by Mister-Hope** [reply]
Emmm, if a user totally fake `X-Forwarded-For` , how can he complete the 3 handshake with server when starting a tcp/ip connection?

Basiclly, in my understanding, most hackers should be faking with `fakeip client proxy1 proxy2` or they are toally faking this field, and connect with a ip they control directly. That ip should be `request.socket.remoteAddress`.

Is the above understanding somewhere uncorrect?

**Reply by Mister-Hope** [reply]
At least, I think we shold have ability, to get the nearst ip the hacker controled sending request.

**Reply by Mister-Hope** [reply]
Trying to explain my understanding:

1. If an attacker is sending request directly, his ip should be `request.socket.remoteAddress`
2. If an attacker is sending a request behind a public or standard proxy, his ip should be added in `X-Forwarded-For` header by proxy, but probably not the first one( that can be faked)
3. If an attacker is  sending a request behind a server he fully controlled, he DO can hide his real ip, but we can still get that server ip and block it.
4. If the user is behind a server reverse proxy( like nginx with php), the above ip should still be in `X-Forwarded-For`
5. If  an attacker is mixing proxy with public and something he controlled, he still can not hide the server ip he controlled, nomatter on the right or on the left of another proxy.

**Reply by lizheming** [reply]
https://github.com/ihackerx/waline-fake-any-ip-poc/blob/main/poc.py#L30 You should notice the code, he uses `True-Client-IP` header to fake ip. And `X-Forwarded-For` has also be changed in this case. It has no trusted IP in the `X-Forwarded-For` header, so check ip at the xff chain whether in the `disallowIPList` is impossible.

```
curl https://vercel-forward-test.vercel.app/ -H 'True-Client-IP: a-fake-ip'
```

![image](https://user-images.githubusercontent.com/424491/152379962-dc088653-1e04-4818-aeeb-b7fd879afb53.png)

**Reply by Mister-Hope** [reply]
In this case what's the value of `request.socket.remoteAddress`（`request.connect.remoteAddress`)? Shouldn't it be the real ip?

He is just making the first one I explain about, "a direct connection", and his true ip should be in the `Request` Object and easy to obtain.

**Reply by lizheming** [reply]
`remoteAddress` is not the real ip because of proxy

![image](https://user-images.githubusercontent.com/424491/152481791-3a51d232-79cf-41c6-a998-65c6c4a2c1f0.png)

**Reply by Mister-Hope** [reply]
So if I am not misunderstanding, the test url is like stuation 4 where I mentioned above, while the proxy is actually doing the "by-pass" job, while simply reads "true-client-ip" and regard it as the real ip then added in the "x-forward-for" header, and hide the actually ip which connects it to the server side?

**Reply by Mister-Hope** [reply]
Shouldn't it be some issues of the "reverse proxy" configuation?

**Reply by Mister-Hope** [reply]
Update:

After speding some time reading the docs, here is my conclusion:

If it's vercel default behavior, then vercel is hiding the connect ip itself, and try to give "the real user ip". Probably with https://nginx.org/en/docs/http/ngx_http_realip_module.html


While the default behavior of nginx should be something like this, since `$remote_addr` should be the proxy ip (or the connection ip by default), `req.connect.remoteAddress` should be actual ip.

```
location / {
    proxy_pass       http://localhost:8000;
    proxy_set_header Host      $host;
    proxy_set_header X-Real-IP $remote_addr;
}
```

If so, that's true there is nothing we can do about it on vercel. But we should be able to block these requests on a self hold env.

**Reply by Mister-Hope** [reply]
cc @ihackerx

**Reply by lizheming** [reply]
In self host mode, we can set `maxIpsCount` to proxy server layers to get a real ip. https://koajs.com/#settings
```
const Waline = require('@waline/vercel');
const Application = new Waline({});
think.app.maxIpsCount = <number>;

module.exports = Application;
```
